### PR TITLE
Cache regexes for scalars -- improves parsing performance

### PR DESCRIPTION
**Pull request by swolchok**
I noticed a lot of time in the scanner was spent futzing with `RegEx`es, so I cached the specific ones that the scanner uses for tokens instead of re-constructing them each time. The two commits combined seem to improve parsing performance by nearly 2x on the sample attached to #158.

### CODE REVIEW

File: src/scanscalar.h [hunk 1]
```diff
@@ -19,7 +19,8 @@ enum FOLD { DONT_FOLD, FOLD_BLOCK, FOLD_FLOW };
 
 struct ScanScalarParams {
   ScanScalarParams()
-      : eatEnd(false),
+      : end(0),
```
Comment (jbeder) [inline src/scanscalar.h]: initialize to nullptr

File: src/scanscalar.h [hunk 2]
```diff
@@ -32,7 +33,8 @@ struct ScanScalarParams {
         leadingSpaces(false) {}
 
   // input:
-  RegEx end;          // what condition ends this scalar?
+  const RegEx* end;   // what condition ends this scalar?
+                      // NOTE: does not have ownership and must not be freed.
```
Comment (jbeder) [inline src/scanscalar.h]: Comment can simply be "unowned" - no need for the detailed info

**Comment by jbeder** [comment]
Nice find! I made a few comments above. I didn't get a chance to test this performance but I will shortly.

File: src/scanscalar.h (hunk 1 above)
Comment (swolchok) [inline src/scanscalar.h]: Oh, didn't realize that C++11 was allowed in yaml-cpp already. Will do!

**Comment by swolchok** [comment]
Anything I can do to help with performance testing?

**Comment by jbeder** [comment]
Sorry, some personal stuff came up so I haven't been able to look at this. Hopefully next week/weekend.

**Comment by swolchok** [comment]
Thanks for the update. Hope your stuff gets resolved smoothly, no rush on my end!

**Comment by jbeder** [comment]
Awesome, pulled as 8c35a8ffabea1c989d9f449bff053081fdc987e1 and 005a6a19ee0af3e284637abe9ecea5501993fc4c (again, cherry-picked).

I added some repetition to the "read" binary (9e37409b4b40e4ca5b026ea9de7ea9784da80ff2) and running

`time util/read -n 100 -c test.yaml`

Before:

```
real 0m26.006s
user 0m25.987s
sys  0m0.010s
```

After:

```
real 0m5.331s
user 0m5.327s
sys  0m0.003s
```

5x improvement!

**Pull request closed without merging**